src/main/resources/stubs/sync/waitgroup.gobra

```diff
@@ -0,0 +1,123 @@
+requires acc(g.WaitGroupP(), dividend/divisor)
+requires !g.WaitMode() && g.UnitDebt(P)
+ensures g.UnitDebt(P) && acc(g.WaitGroupStarted(), dividend/divisor)
+func (g *WaitGroup) Start(ghost dividend int, ghost divisor int, ghost P pred())
+
+ghost
+requires dividendP >= 0 && divisorP > 0 // p >= 0
+requires dividendQ > 0 && divisorQ > 0 // q > 0
+requires dividendP * divisorQ + dividendQ * divisorP == divisorP * divisorQ // p + q == 1
+requires acc(g.WaitGroupP(), dividendP/divisorP)
+requires acc(g.WaitGroupStarted(), dividendQ/divisorQ)
+ensures g.WaitGroupP() && g.WaitMode()
+func (g *WaitGroup) SetWaitMode(ghost dividendP int, ghost divisorP int, ghost dividendQ int, ghost divisorQ int)
+
+// Simplified version of the debt redistribution rule. The most general version cannot be written in Gobra
+// without support for magic wands
+ghost
+requires P() && g.UnitDebt(P)
+ensures g.UnitDebt(pred_true!<!>)
+func (g *WaitGroup) PayDebt(ghost P pred())
+
+// Simplified version of the debt redistribution rule, instantiated with P == { pred_true!<!> } and Q == { R }.
+// This is the only rule that generates a Token.
+ghost
+requires R() && g.UnitDebt(pred_true!<!>)
+ensures g.UnitDebt(R) && g.Token(R)
+func (g *WaitGroup) GenerateToken(ghost R pred())
+
+requires dividend >= 0 && divisor > 0 // p >= 0
+requires n > 0 && dividend > 0 ==> acc(g.WaitGroupP(), dividend/divisor) && !g.WaitMode()
+requires (n > 0 && dividend == 0) ==> g.UnitDebt(P)
+requires n < 0 ==> acc(g.UnitDebt(pred_true!<!>), -n/1)
+ensures (n > 0 && dividend > 0) ==> acc(g.WaitGroupP(), dividend/divisor)
+ensures (n > 0 && dividend == 0) ==> g.UnitDebt(P)
+ensures n > 0 ==> acc(g.UnitDebt(pred_true!<!>), n/1)
+// this is actually necessary, otherwise Gobra cannot prove that Add does not modify the wait mode
+ensures (n > 0 && dividend > 0) ==> g.WaitMode() == old(g.WaitMode())
+func (g *WaitGroup) Add(n int, ghost dividend int, ghost divisor int, ghost P pred())
+
+requires g.UnitDebt(pred_true!<!>)
+func (g *WaitGroup) Done()
+
+// `dividend > 0 && divisor > 0` implies that the permission p in Martin's proposal is greater than 0
+requires dividend > 0 && divisor > 0 // p > 0
+requires acc(g.WaitGroupP(), dividend/divisor)
+requires g.WaitMode()
+requires acc(g.Token(P))
+ensures P()
+ensures acc(g.WaitGroupP(), dividend/divisor)
+func (g *WaitGroup) Wait(ghost dividend int, ghost divisor int, ghost P pred())
+
+pred CollectFractions(ghost P seq[pred()], ghost dividends seq[int], ghost divisors seq[int]) {
+	len(P) == len(dividends) && len(dividends) == len(divisors) &&
+	forall i int :: 0 <= i && i < len(P) ==> dividends[i] >= 0 && divisors[i] > 0 && acc(P[i](), dividends[i]/divisors[i])
+}
+
+// Problem: CollectFractions(P, [x1, ..., xn], [y1, ..., yn]) does not imply CollectFractions(P, [x1', ..., xn'], [y1', ..., yn']) even if for all i, xi / yi == xi' / yi'. This will not be a problem once we add
+// the Perm type to Gobra. Meanwhile, a new viewshift is required for this case.
```

Comment:
What is the status of this viewshift? Does it already exist?

Reply:
I did not implement it as it will become completely unnecessary after we introduce the Perm Type and no example that I came accross thus far needed this. However, if you find it useful, I will write it :) stay tuned for the next commits

Reply:
Makes sense. I was just not sure what ‚a new viewshift is required‘ exactly means

Reply:
I added the view shift in the last commit. It is totally a bit of an ugly thing. I think that, if time allows, we should focus on adding the Perm type to the language